go/vt/schemadiff/table.go

```diff
@@ -163,8 +162,159 @@ func NewCreateTableEntity(c *sqlparser.CreateTable) *CreateTableEntity {
 
 // normalize normalizes table definition:
 // - setting names to all keys
-func (c *CreateTableEntity) normalize() {
-	// let's verify all keys have names
+// - table option case (upper/lower/special)
+// The function returns this receiver as courtesy
+func (c *CreateTableEntity) normalize() *CreateTableEntity {
+	c.normalizeUnnamedKeys()
+	c.normalizeTableOptions()
+	c.normalizeColumnOptions()
+	c.normalizePartitionOptions()
+	return c
+}
+
+func (c *CreateTableEntity) normalizeTableOptions() {
+	for _, opt := range c.CreateTable.TableSpec.Options {
+		switch strings.ToUpper(opt.Name) {
+		case "CHARSET", "COLLATE":
+			opt.String = strings.ToLower(opt.String)
+			if charset, ok := charsetAliases[opt.String]; ok {
+				opt.String = charset
+			}
+		case "ENGINE":
+			opt.String = strings.ToUpper(opt.String)
+			if engineName, ok := engineCasing[opt.String]; ok {
+				opt.String = engineName
+			}
+		case "ROW_FORMAT":
+			opt.String = strings.ToUpper(opt.String)
+		}
+	}
+}
+
+func (c *CreateTableEntity) normalizeColumnOptions() {
+	tableCharset := defaultCharset
+	tableCollation := ""
+	for _, option := range c.CreateTable.TableSpec.Options {
+		switch strings.ToUpper(option.Name) {
+		case "CHARSET":
+			tableCharset = option.String
+		case "COLLATE":
+			tableCollation = option.String
+		}
+	}
+	if tableCollation == "" {
+		tableCollation = defaultCollations[tableCharset]
+	}
+
+	for _, col := range c.CreateTable.TableSpec.Columns {
+		if col.Type.Options == nil {
+			col.Type.Options = &sqlparser.ColumnTypeOptions{}
+		}
+
+		// Map known lowercase fields to always be lowercase
+		col.Type.Type = strings.ToLower(col.Type.Type)
+		col.Type.Charset = strings.ToLower(col.Type.Charset)
+		col.Type.Options.Collate = strings.ToLower(col.Type.Options.Collate)
+
+		// See https://dev.mysql.com/doc/refman/8.0/en/create-table.html
+		// If neither NULL nor NOT NULL is specified, the column is treated as though NULL had been specified.
+		// That documentation though is not 100% true. There's an exception, and that is
+		// the `explicit_defaults_for_timestamp` flag. When that is disabled (the default on 5.7),
+		// a timestamp defaults to `NOT NULL`.
+		//
+		// We opt here to instead remove that difference and always then add `NULL` and treat
+		// `explicit_defaults_for_timestamp` as always enabled in the context of DDL for diffing.
```

Comment:
This is another case where the MySQL 8.0 behavior is considered the standard and the behavior we assume.